ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+						modifiedAccounts[addr] = nil
+						continue
+					}
+
+					modifiedAccounts[addr] = accountDataToOnline(addr, &d, genesisProto)
+				}
+			}
+		}
+
+		ao.accountsMu.RUnlock()
+
+		// Build up a set of candidate accounts.  Start by loading the
+		// top N + len(modifiedAccounts) accounts from disk (accountdb).
+		// This ensures that, even if the worst case if all in-memory
+		// changes are deleting the top accounts in accountdb, we still
+		// will have top N left.
+		//
+		// Keep asking for more accounts until we get the desired number,
+		// or there are no more accounts left.
+		candidates := make(map[basics.Address]*ledgercore.OnlineAccount)
+		batchOffset := uint64(0)
+		batchSize := uint64(1024)
+		var dbRound basics.Round
+		for uint64(len(candidates)) < n+uint64(len(modifiedAccounts)) {
+			var accts map[basics.Address]*ledgercore.OnlineAccount
+			start := time.Now()
+			ledgerAccountsonlinetopCount.Inc(nil)
+			err = ao.dbs.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+				accts, err = accountsOnlineTop(tx, rnd, batchOffset, batchSize, genesisProto)
+				if err != nil {
+					return
+				}
+				dbRound, err = accountsRound(tx)
+				return
+			})
+			ledgerAccountsonlinetopMicros.AddMicrosecondsSince(start, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			if dbRound != currentDbRound {
+				break
+			}
+
+			for addr, data := range accts {
+				if !(data.VoteFirstValid <= voteRnd && voteRnd <= data.VoteLastValid) {
+					continue
+				}
+				candidates[addr] = data
+			}
+
+			// If we got fewer than batchSize accounts, there are no
+			// more accounts to look at.
+			if uint64(len(accts)) < batchSize {
+				break
+			}
+
+			batchOffset += batchSize
+		}
+		if dbRound != currentDbRound && dbRound != basics.Round(0) {
+			// database round doesn't match the last au.dbRound we sampled.
+			ao.accountsMu.RLock()
+			for currentDbRound >= ao.cachedDBRoundOnline && currentDeltaLen == len(ao.deltas) {
+				ao.accountsReadCond.Wait()
+			}
+			continue
+		}
+
+		// Now update the candidates based on the in-memory deltas.
+		for addr, oa := range modifiedAccounts {
+			if oa == nil {
+				delete(candidates, addr)
+			} else {
+				candidates[addr] = oa
+			}
+		}
+
+		// Get the top N accounts from the candidate set, by inserting all of
+		// the accounts into a heap and then pulling out N elements from the
+		// heap.
+		topHeap := &onlineTopHeap{
+			accts: nil,
+		}
+
+		for _, data := range candidates {
+			heap.Push(topHeap, data)
+		}
+
+		var res []*ledgercore.OnlineAccount
+		for topHeap.Len() > 0 && uint64(len(res)) < n {
+			acct := heap.Pop(topHeap).(*ledgercore.OnlineAccount)
+			res = append(res, acct)
+		}
+
+		return res, nil
+	}
+}
+
+var ledgerAccountsonlinetopCount = metrics.NewCounter("ledger_accountsonlinetop_count", "calls")
+var ledgerAccountsonlinetopMicros = metrics.NewCounter("ledger_accountsonlinetop_micros", "µs spent")
```

Comment:
nit: unless there's precedent for this, I think it's better to put these at the top of the file

Reply:
IDK why, but this is how all the metrics in the ledger package are defined